### ISSUE TITLE
lgb.importance: Error in data.table::rbindlist(l = list(single_tree_dt, tree_node_leaf[c("leaf_index", : Unsupported type 'NULL'

### ISSUE THREAD

**Issue by wangshuailong** [Microsoft/LightGBM#859]
## Environment info
Operating System: Mac OS Sierra 10.12.6
CPU: 2.5Ghz Intel i7
C++/Python/R version: Microsoft R Open 3.4.0

## Error Message: 
when I use lgb.importance

> Error in data.table::rbindlist(l = list(single_tree_dt, tree_node_leaf[c("leaf_index",  : 
>   Unsupported type 'NULL'

## Reproducible examples

> data(agaricus.train, package = "lightgbm")
> data(agaricus.test, package = "lightgbm")
> train <- agaricus.train
> test <- agaricus.test
> 
> bst <- lightgbm(data = train$data,
>                 label = train$label,
>                 num_leaves = 4,
>                 learning_rate = 1,
>                 nrounds = 2,
>                 objective = "binary")
> 
> bst_importance <- lgb.importance(bst)

The error produced when I dealt with my own data, but also can be reproducible with this one.

## Steps to reproduce
As Above

## SessonInfo

> R version 3.4.0 (2017-04-21)
> Platform: x86_64-apple-darwin15.6.0 (64-bit)
> Running under: macOS Sierra 10.12.6
> 
> Matrix products: default
> BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
> LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
> 
> locale:
> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
> 
> attached base packages:
> [1] stats     graphics  grDevices utils     datasets 
> [6] methods   base     
> 
> other attached packages:
>  [1] stringr_1.2.0   lightgbm_2.0.5  R6_2.2.0       
>  [4] ggplot2_2.2.1   hydroGOF_0.3-8  zoo_1.8-0      
>  [7] magrittr_1.5    xgboost_0.6-4   lubridate_1.6.0
> [10] tidyr_0.6.1     Matrix_1.2-10   dplyr_0.5.0    
> [13] readr_1.1.0    
> 
> loaded via a namespace (and not attached):
>  [1] automap_1.0-14    Rcpp_0.12.10     
>  [3] compiler_3.4.0    plyr_1.8.4       
>  [5] class_7.3-14      xts_0.9-7        
>  [7] tools_3.4.0       gstat_1.1-5      
>  [9] jsonlite_1.4      tibble_1.3.0     
> [11] gtable_0.2.0      lattice_0.20-35  
> [13] DBI_0.6-1         e1071_1.6-8      
> [15] RevoUtils_10.0.4  hms_0.3          
> [17] grid_3.4.0        spacetime_1.2-0  
> [19] reshape_0.8.6     data.table_1.10.4
> [21] sp_1.2-4          scales_0.4.1     
> [23] intervals_0.15.1  assertthat_0.2.0 
> [25] colorspace_1.3-2  stringi_1.1.5    
> [27] lazyeval_0.2.0    munsell_0.4.3    
> [29] hydroTSM_0.4-2-1  FNN_1.1
> 

## Other Important Info
I had an error with lgb.Dataset as follow

> Error in lgb.call("LGBM_DatasetCreateFromMat_R", ret = handle, private$raw_data, : api error: cannot open data file 2016

I followed advise from somewhere to reinstall my lightgbm. It's quite tricky this time because of the C compiler. I installed from the source (but not install_github), but it was installed in the wrong path (I don't know why), so then I copied it into the library path.

**Comment by guolinke** [comment]
the first error is caused by single-leaf tree, which only have constant output (don't  have `leaf_index` field). @Laurae2 can you help to fix this ? 

And it seems there are many strange issues for the R-package in OSX. 
How did you install the lightgbm R ? are there any log ?

**Comment by wangshuailong** [comment]
Thank you for your help. 
At first I installed by devtool::install_github, but it seems it was compiled by Apple Clang. It didn't work, and I had no idea how to ask it to compile with Gcc. So i gave up. 
Then I installed as the document. 
git clone --recursive https://github.com/Microsoft/LightGBM
cd LightGBM/R-package
# export CXX=g++-7 CC=gcc-7 # for OSX
R CMD INSTALL --build . --no-multiarch

I saw a folder "LightGBM" under the library dictionary. But once I load it in Rstudio. There was error message: 
There is no package named LightGBM 
Something like this. 
I noticed the folder tree looked like LightGBM/R-package/.... I thought maybe what was under the R-Package should be under the upper level folder LightGBM... this caused the error?
I removed it and tried many times. Once it was installed under ~/R/library. It's not my R libpath, but i thought the organization of the files was what it should be, so i copied it to the right libpath. It works, but has this error.

Sorry I'm not good at software engineer... so I might make some obvious mistakes...

Not good edit from phone... 

thank you.

**Comment by guolinke** [comment]
@wangshuailong 
you should uncomment this line in OXS: ``# export CXX=g++-7 CC=gcc-7 # for OSX`` .

```
git clone --recursive https://github.com/Microsoft/LightGBM
cd LightGBM/R-package
# export CXX=g++-7 CC=gcc-7 # for OSX
R CMD INSTALL --build . --no-multiarch
```

**Comment by wangshuailong** [comment]
Yes I uncommitted these lines. Didn't write correctly.

**Comment by guolinke** [comment]
@wangshuailong 
did you install gcc ? 
```
brew install cmake
brew install gcc --without-multilib
```

**Comment by Laurae2** [comment]
@guolinke see PR #862 for `lgb.importance` fix

**Comment by wangshuailong** [comment]
@guolinke @Laurae2 Thank you so much! It's working Now!
